src/sage/modules/vector_space_morphism.py

```diff
@@ -744,7 +781,22 @@ def linear_transformation(arg0, arg1=None, arg2=None, side='left'):
         if side == 'right':
             arg2 = arg2.transpose()
     elif isinstance(arg2, (list, tuple)):
-        pass
+        # For CombinatorialFreeModule and similar vector spaces with basis keys,
+        # we need to ensure the list elements are properly interpreted as
+        # vectors in the codomain. Convert to matrix representation early
+        # to avoid issues in VectorSpaceHomspace.__call__
+        if len(arg2) != D.dimension():
+            msg = "number of images should equal the size of the domain's basis (={0}), not {1}"
+            raise ValueError(msg.format(D.dimension(), len(arg2)))
```

Comment:
```suggestion
            raise ValueError(f"number of images should equal the dimension of the domain (={D.dimension()}), not {len(arg2)}")
```
Not sure, but I think that `f"..."` is now preferred within sage.